crypto/crypto.go

```diff
@@ -6,112 +6,120 @@ import (
 // signatures.
-func New(impl consensus.CryptoImpl) consensus.Crypto {
-	return base{CryptoImpl: impl}
+func New(impl consensus.CryptoBase) consensus.Crypto {
+	return &crypto{CryptoBase: impl}
 }
 
 // InitConsensusModule gives the module a reference to the Modules object.
 // It also allows the module to set module options using the OptionsBuilder.
-func (base base) InitConsensusModule(mods *consensus.Modules, cfg *consensus.OptionsBuilder) {
-	if mod, ok := base.CryptoImpl.(consensus.Module); ok {
+func (c *crypto) InitConsensusModule(mods *consensus.Modules, cfg *consensus.OptionsBuilder) {
+	c.mods = mods
+	if mod, ok := c.CryptoBase.(consensus.Module); ok {
 		mod.InitConsensusModule(mods, cfg)
 	}
 }
 
 // CreatePartialCert signs a single block and returns the partial certificate.
-func (base base) CreatePartialCert(block *consensus.Block) (cert consensus.PartialCert, err error) {
-	sig, err := base.Sign(block.Hash())
+func (c crypto) CreatePartialCert(block *consensus.Block) (cert consensus.PartialCert, err error) {
+	sig, err := c.Sign(block.ToBytes())
 	if err != nil {
 		return consensus.PartialCert{}, err
 	}
 	return consensus.NewPartialCert(sig, block.Hash()), nil
 }
 
 // CreateQuorumCert creates a quorum certificate from a list of partial certificates.
-func (base base) CreateQuorumCert(block *consensus.Block, signatures []consensus.PartialCert) (cert consensus.QuorumCert, err error) {
+func (c crypto) CreateQuorumCert(block *consensus.Block, signatures []consensus.PartialCert) (cert consensus.QuorumCert, err error) {
 	// genesis QC is always valid.
 	if block.Hash() == consensus.GetGenesis().Hash() {
 		return consensus.NewQuorumCert(nil, 0, consensus.GetGenesis().Hash()), nil
 	}
-	sigs := make([]consensus.Signature, 0, len(signatures))
+	sigs := make([]consensus.QuorumSignature, 0, len(signatures))
 	for _, sig := range signatures {
 		sigs = append(sigs, sig.Signature())
 	}
-	sig, err := base.CreateThresholdSignature(sigs, block.Hash())
+	sig, err := c.Combine(sigs...)
 	if err != nil {
 		return consensus.QuorumCert{}, err
 	}
 	return consensus.NewQuorumCert(sig, block.View(), block.Hash()), nil
 }
 
 // CreateTimeoutCert creates a timeout certificate from a list of timeout messages.
-func (base base) CreateTimeoutCert(view consensus.View, timeouts []consensus.TimeoutMsg) (cert consensus.TimeoutCert, err error) {
+func (c crypto) CreateTimeoutCert(view consensus.View, timeouts []consensus.TimeoutMsg) (cert consensus.TimeoutCert, err error) {
 	// view 0 is always valid.
 	if view == 0 {
 		return consensus.NewTimeoutCert(nil, 0), nil
 	}
-	sigs := make([]consensus.Signature, 0, len(timeouts))
+	sigs := make([]consensus.QuorumSignature, 0, len(timeouts))
 	for _, timeout := range timeouts {
 		sigs = append(sigs, timeout.ViewSignature)
 	}
-	sig, err := base.CreateThresholdSignature(sigs, view.ToHash())
+	sig, err := c.Combine(sigs...)
 	if err != nil {
 		return consensus.TimeoutCert{}, err
```

Comment:
We should avoid returning empty structs when we have error; it causes unnecessary allocation.